include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -936,6 +929,22 @@ inline void SessionImpl<T>::Run(const RunOptions& run_options, const IoBinding&
   ThrowOnError(GetApi().RunWithBinding(this->p_, run_options, io_binding));
 }
 
+inline void CallbackBridge(void* user_data, OrtValue** outputs, size_t num_outputs, OrtStatusPtr status) {
+  RunAsyncCallbackStdFn* callback = reinterpret_cast<RunAsyncCallbackStdFn*>(user_data);
+  std::vector<Ort::Value> output_values;
+  for (size_t ith = 0; ith < num_outputs; ++ith) {
+    output_values.emplace_back(outputs[ith]);
+  }
+  (*callback)(output_values, Ort::Status{status});
+}
+
+template <typename T>
+inline void SessionImpl<T>::RunAsync(const RunOptions& run_options, const char* const* input_names, const Value* input_values, size_t input_count,
+                                     const char* const* output_names, size_t output_count, RunAsyncCallbackStdFn& callback) {
+  auto ort_input_values = reinterpret_cast<const OrtValue* const*>(input_values);
+  ThrowOnError(GetApi().RunAsync(this->p_, run_options, input_names, ort_input_values, input_count, output_names, output_count, CallbackBridge, &callback));
```

Comment:
`callback` may go out of its scope and destructed before being used. maybe pass `new CPP_CALLBACK(std::move(callback))` and in implementation, use `std::unique_ptr<CPP_CALLBACK> cb(reinterpret_cast<CPP_CALLBACK*>(state));` to make sure memory get managed properly.

Reply:
We could pass the std func by value.
However, life won't be easier a lot since customers still have to make sure all the captures are alive.
They should be allowed to have the illusion that their callback function could die before being called.

Reply:
Passing by value helps with any expectation management with users and keeps things simple + avoid any new dynamic memory allocation.

Reply:
As per discussion let's stick with a vanilla function ptr based callback in the C++ API as well. We can add more bells and whistles on top it later as it doesn't affect the ABI.